node/network/statement-distribution/src/lib.rs

```diff
@@ -0,0 +1,1376 @@
+	/// per validator. These will be the first ones we see. The statement is assumed
+	/// to have been checked, including that the validator index is not out-of-bounds and
+	/// the signature is valid.
+	///
+	/// Any other statements or those that reference a candidate we are not aware of cannot be accepted.
+	fn note_statement(&mut self, statement: SignedFullStatement) -> NotedStatement {
+		let validator_index = statement.validator_index();
+		let comparator = StoredStatementComparator {
+			compact: statement.payload().to_compact(),
+			validator_index,
+			signature: statement.signature().clone(),
+		};
+
+		let stored = StoredStatement {
+			comparator: comparator.clone(),
+			statement,
+		};
+
+		match comparator.compact {
+			CompactStatement::Candidate(h) => {
+				let seconded_so_far = self.seconded_counts.entry(validator_index).or_insert(0);
+				if *seconded_so_far >= 2 {
+					return NotedStatement::NotUseful;
+				}
+
+				self.candidates.insert(h);
+				if self.seconded_statements.insert(stored) {
+					*seconded_so_far += 1;
+
+					// This will always return `Some` because it was just inserted.
+					NotedStatement::Fresh(self.seconded_statements.get(&comparator)
+						.expect("Statement was just inserted; qed"))
+				} else {
+					NotedStatement::UsefulButKnown
+				}
+			}
+			CompactStatement::Valid(h) | CompactStatement::Invalid(h) => {
+				if !self.candidates.contains(&h) {
+					return NotedStatement::NotUseful;
+				}
+
+				if self.other_statements.insert(stored) {
+					// This will always return `Some` because it was just inserted.
+					NotedStatement::Fresh(self.other_statements.get(&comparator)
+						.expect("Statement was just inserted; qed"))
+				} else {
+					NotedStatement::UsefulButKnown
+				}
+			}
+		}
+	}
+
+	/// Get an iterator over all statements for the active head. Seconded statements come first.
+	fn statements(&self) -> impl Iterator<Item = &'_ StoredStatement> + '_ {
+		self.seconded_statements.iter().chain(self.other_statements.iter())
+	}
+
+	/// Get an iterator over all statements for the active head that are for a particular candidate.
+	fn statements_about(&self, candidate_hash: Hash)
+		-> impl Iterator<Item = &'_ StoredStatement> + '_
+	{
+		self.statements().filter(move |s| s.compact().candidate_hash() == &candidate_hash)
+	}
+}
+
+/// Check a statement signature under this parent hash.
+fn check_statement_signature(
+	head: &ActiveHeadData,
+	relay_parent: Hash,
+	statement: &SignedFullStatement,
+) -> Result<(), ()> {
+	let signing_context = SigningContext {
+		session_index: head.session_index,
+		parent_hash: relay_parent,
+	};
+
+	match head.validators.get(statement.validator_index() as usize) {
+		None => return Err(()),
+		Some(v) => statement.check_signature(&signing_context, v),
+	}
+}
+
+#[derive(Encode, Decode)]
+enum WireMessage {
+	/// relay-parent, full statement.
+	Statement(Hash, SignedFullStatement),
```

Comment:
will this be extended in the future?

Reply:
Possible, but unlikely. I'm not sure how to handle protocol version changes with the network bridge, but it's possible we would introduce a new message enum for those.

Reply:
I added a `#[codec(index = "0")]` for future-proofing.